docs/interfaces.md

```diff
@@ -0,0 +1,431 @@
+ChaincodeInvoke(transactionProposal []byte, ctx interface{}) (response, error)
+```
+
+```go
+// ECC must trigger this when sealedState is available using load() and ecc starts; this is not really a public interface
+createEnclaveWithSealedState(sealedState []byte) error
+
+// Storage interface
+// TODO how to store sealedState, where and under what identifier (e.g. chaincode_id);
+load(chaincode_id string) (sealedState []byte, err error)
+store(chaincode_id string, sealedState []byte) error
+```
+
+This interface is implemented by ECC to let a chaincode enclave call into the peer
+
+```c++
+
+// logging
+void ocall_print_string([in, string] const char *str);
+
+
+// state access
+// TBD does this key already contain the chaincode_id/namespace?
+void ocall_get_state(
+        [in, string] const char *key, 
+        [out, size=max_val_len] uint8_t *val, uint32_t max_val_len, [out] uint32_t *val_len,
+        [in, out] sgx_cmac_128bit_tag_t *cmac,
+        [user_check] void *u_shim_ctx);
+
+void ocall_put_state(
+        [in, string] const char *key,
+        [in, size=val_len] uint8_t *val, uint32_t val_len,
+        [user_check] void *u_shim_ctx);
+
+void ocall_get_state_by_partial_composite_key(
+        [in, string] const char *comp_key,
+        [out, size=max_len] uint8_t *values, uint32_t max_len, [out] uint32_t *values_len,
+        [in, out] sgx_cmac_128bit_tag_t *cmac,
+        [user_check] void *u_shim_ctx);
+
+
+// verify msp with the help of tlcc
+void ocall_verify_msp_id(
+        [in, string] const char *msp_id,
+        [in, size=32] uint8_t *nonce,
+        [out] bool isValid, 
+        [out] sgx_cmac_128bit_tag_t *cmac);
+
+
+// attestation
+void ocall_get_spid(); // TBD: is spid only needed for EPID?
+
+// creates an attestation evidence; the evidence_request includes all the form of evidence (EPID/DCAP), the quote, and additional information required to complete the evidence request
+void ocall_create_evidence(
+        [in] evidence_request
+        [out] evidence
+);
+```
+
+## State:
+An ECC instance keeps a local reference of the FPC chaincode enclave id.
+
+```go
+type EnclaveChaincode struct {
+        erccStub ercc.EnclaveRegistryStub
+        tlccStub tlcc.TLCCStub
+        enclave  enclave.Stub
+        verifier crypto.Verifier
+        ev       attestation.EvidenceService // abstracts evidence creation for EPID/DCAP
+}
+
+type StubImpl struct {
+	eid C.enclave_id_t
+}
+```
+
+
+# ECC_Enclave
+
+## Interface:
+The ECC_Enclave interface specifies interface of an FPC chaincode enclave. 
+
+Note: `sealed credentials` is renamed to `sealed state`
+
+```c++
+// creates an new enclave
+public int ecall_create_enclave(
+        [in, size=cc_params_len] uint8_t *cc_params, uint32_t cc_params_len,
+        [out] uint8_t *credentials,     // TBD is this a fixed length type?
+        [out, size=max_sealed_state_len] uint8_t *sealed_state, uint32_t max_sealed_state_len,
+        [out] uint32_t *sealed_data_len);
```

Comment:
It's a matter of taste but i think it would be nice to have a separate library/module for sealed state-management and don't put it here in these ecalls.  The nice part i think would be that we can start without sealed state and when adding sealed state none of the main ecall/ocalls here change but there will be this additional library with it's own edl and decisions like having separate write-once sealed blobs vs having single updateable blobs  can be all defined separatedly?
Admittedly, though, i guess the current UML seem to specify it this way, so we might have to fix it. But there seemt to be other small consistency issues with the diagrams which we might have to fix anyway at some point?